toolchain/check/generic_region_stack.h

```diff
@@ -60,6 +60,60 @@ class GenericRegionStack {
     pending_generic_ids_.back().generic_id = generic_id;
   }
 
+  // While held alive, instructions added as dependents or to the eval block are
+  // treated as part of a single commit. If `Discard` is called, then those
+  // instructions will all be dropped from the GenericRegionStack. Otherwise,
+  // they will be kept.
+  class [[nodiscard]] Commit {
+    friend GenericRegionStack;
+
+   public:
+    ~Commit() { std::move(*this).Keep(); }
+
+    auto Discard() && -> void {
+      if (stack_) {
+        stack_->dependent_inst_stack_.PopArray();
+        stack_->pending_eval_block_stack_.PopArray();
+        stack_->ongoing_commit_ = false;
+        stack_ = nullptr;
+      }
```

Comment:
I think we'll need to do something with the `ConstantsInGenericMap` too -- either removing the entries from there corresponding to the things we removed from the eval block, or re-adding the items in that map back into the eval block if they're referenced again.

Specifically, if a generic references constant X during a transaction that gets reverted, and then later references the constant X again, if it's still in the `ConstantsInGenericMap` then I think we won't re-add it to the eval block, but we presumably need to.